src/pudl/analysis/record_linkage/link_cross_year.py

```diff
@@ -0,0 +1,273 @@
+            average_dist_matrix[i, j] = average_dist
+            average_dist_matrix[j, i] = average_dist
+
+    return average_dist_matrix
+
+
+@op
+def compute_distance_with_year_penalty(
+    config: PenalizeReportYearDistanceConfig,
+    feature_matrix: FeatureMatrix,
+    original_df: pd.DataFrame,
+) -> DistanceMatrix:
+    """Compute a distance matrix and penalize records from the same year."""
+    return DistanceMatrix(feature_matrix.matrix, original_df, config)
+
+
+class DBSCANConfig(Config):
+    """Configuration for DBSCAN step."""
+
+    #: See :class:`sklearn.cluster.DBSCAN` for details.
+    eps: float = 0.1
+    min_samples: int = 2
+
+
+@op
+def cluster_records_dbscan(
+    config: DBSCANConfig, distance_matrix: DistanceMatrix, original_df: pd.DataFrame
+) -> pd.DataFrame:
+    """Generate initial IDs using DBSCAN algorithm."""
+    # DBSCAN is very efficient when passed a sparse radius neighbor graph
+    neighbor_computer = NearestNeighbors(radius=config.eps, metric="precomputed")
+    neighbor_computer.fit(distance_matrix.distance_matrix)
+    neighbor_graph = neighbor_computer.radius_neighbors_graph(mode="distance")
+
+    # Classify records
+    classifier = DBSCAN(metric="precomputed", eps=config.eps, min_samples=2)
+
+    # Create dataframe containing only report year and label columns
+    return pd.DataFrame(
+        {
+            "report_year": original_df.loc[:, "report_year"],
+            "record_label": classifier.fit_predict(neighbor_graph),
+        }
+    )
+
+
+class SplitClustersConfig(Config):
+    """Configuration for AgglomerativeClustering used to split overmerged clusters."""
+
+    #: See :class:`sklearn.cluster.AgglomerativeClustering` for details.
+    distance_threshold: float = 0.3
+
+
+@op
+def split_clusters(
+    config: SplitClustersConfig,
+    distance_matrix: DistanceMatrix,
+    id_year_df: pd.DataFrame,
+) -> pd.DataFrame:
+    """Split clusters with multiple records from same report_year.
+
+    DBSCAN will sometimes match records from the same report year, which breaks the
+    assumption that there should only be one record for each entity from a single
+    report year. To fix this, agglomerative clustering will be applied to each
+    such cluster. Agglomerative clustering could replace DBSCAN in the initial linkage
+    step to avoid these matches in the first place, however, it is very inneficient on
+    a large number of records, so applying to smaller sets of overmerged records is
+    much faster and uses much less memory.
+    """
+
+    def _generate_cluster_ids(max_cluster_id: int) -> int:
+        """Get new unique cluster id."""
+        while True:
+            max_cluster_id += 1
+            yield max_cluster_id
+
+    cluster_id_generator = _generate_cluster_ids(id_year_df.record_label.max())
+    classifier = AgglomerativeClustering(
+        metric="precomputed",
+        linkage="average",
+        distance_threshold=config.distance_threshold,
```

Comment:
Yeah `eps` and `distance_threshold` are definitely not identical, however they both do apply an upper bound on the distances within a cluster, so I agree it would make sense to make them the same by default.